lib/apis/moray.js

```diff
@@ -631,35 +660,121 @@ Moray.prototype._addFilterOptions = function (params) {
         }
 
         if (splitted.length == 2) {
-            order = splitted[1] || defaultOrder;
+            order = splitted[1] || common.DEFAULT_SORT_ORDER;
         }
 
-        options.sort = {
+        sortOptions.push({
             attribute: property,
             order: order
-        };
-    } else {
-        options.sort = {
-            attribute: '_id',
-            order: 'ASC'
-        };
+        });
     }
 
-    if (params.offset) {
-        options.offset = params.offset;
+    // If the sorting options will not output a totally ordered results set,
+    // add a sort option that will. This makes the last item of any
+    // results set (even when not using a marker, for example when getting
+    // the first page of paginated results) usable as a marker for getting
+    // the next page of paginated results.
+    var sortOptionsHasStrictTotalOrder =
+        sortOptions.some(function (sortOption) {
+            return common.isStrictTotalOrderField(sortOption.attribute);
+        });
+
+    if (!sortOptionsHasStrictTotalOrder) {
+        sortOptions.push({
+            attribute: common.strictTotalOrderField(),
+            order: common.DEFAULT_SORT_ORDER
+        });
     }
 
-    if (params.limit) {
-        options.limit = params.limit;
+    return sortOptions;
+}
+
+/*
+ * Build the appropriate LDAP filters for the marker "marker"
+ * and the sort options "sortOptions". Returns a list of strings
+ * representing ldap filters. An empty list is returned if not filters
+ * needs to be used for the corresponding marker and sortOptions.
+ */
+function _buildFiltersFromMarker(marker, sortOptions) {
+    assert.object(marker, 'marker must be an object');
+    assert.object(sortOptions, 'sortOptions must be an object');
+
+    var markerFilters = [];
+
+    var sortOptionsLookup = {};
+    sortOptions.forEach(function (sortOption) {
+        sortOptionsLookup[sortOption.attribute] = sortOption.order;
+    });
+
+    Object.keys(marker).forEach(function (markerKey) {
+        var markerValue = marker[markerKey];
+        var order;
+        var filter;
+
+        if (markerValue !== null) {
+            order = sortOptionsLookup[markerKey];
+            if (order === undefined)
+                order = common.DEFAULT_SORT_ORDER;
+
+            if (common.isSortOrderDesc(order))
+                filter = PARAM_FILTER_LE;
+            else
+                filter = PARAM_FILTER_GE;
+
+            filter = sprintf(filter, markerKey, markerValue);
+            markerFilters.push(filter);
```

Comment:
It feels like we should probably be using something like the
["ldapjs" Filters API][1] to construct these filters, rather than
a bunch of string and array manipulation.  That API is capable
of parsing filter strings, composing trees of filter objects,
and rendering those trees back to a string, etc.

[1]: http://ldapjs.org/filters.html